### PR TITLE
api: add api support for Update events

### LINKED ISSUE #195: Updating of additonal config field of OBC
I am not sure whether bug or current design limitation. In rook quota for OBC is supported with help of `additionalConfig` string Map. When I try to update the field it is not working via `kubectl edit` or `kubectl apply` the request is not handled properly.

I can see following logs from rook-operator
```console
I1014 10:05:53.347471       6 controller.go:210]  "msg"="reconciling claim" "key"="default/ceph-delete-bucket" 
I1014 10:05:53.347535       6 helpers.go:80]  "msg"="getting claim for key" "key"="default/ceph-delete-bucket" 
I1014 10:05:53.349255       6 helpers.go:181]  "msg"="getting ObjectBucketClaim's StorageClass" "key"="default/ceph-delete-bucket" 
I1014 10:05:53.353931       6 helpers.go:186]  "msg"="got StorageClass" "key"="default/ceph-delete-bucket" "name"="rook-ceph-delete-bucket"
I1014 10:05:53.354343       6 helpers.go:63]  "msg"="checking OBC for OB name, this indicates provisioning is complete" "key"="default/ceph-delete-bucket" "ceph-delete-bucket"=null
I1014 10:05:53.354404       6 helpers.go:65]  "msg"="provisioning already completed" "key"="default/ceph-delete-bucket" "ObjectBucket"="obc-default-ceph-delete-bucket"
I1014 10:05:53.354424       6 controller.go:246]  "msg"="skipping provision" "key"="default/ceph-delete-bucket"
```

If I understand correctly current code follow in `lib-bucket-provisioner` the request reaches `syncHandler`. 

```go
   // ***********************
    // Delete or Revoke Bucket
    // ***********************
    if obc.ObjectMeta.DeletionTimestamp != nil {
        log.Info("OBC deleted, proceeding with cleanup")
        return c.handleDeleteClaim(key, obc)
    }

    // *******************************************************
    // Provision New Bucket or Grant Access to Existing Bucket
    // *******************************************************
    if !shouldProvision(obc) {
        log.Info("skipping provision")
        return nil
    }

    // update the OBC's status to pending before any provisioning related errors can occur
    obc, err = updateObjectBucketClaimPhase(
        c.libClientset,
```
Firsts it check for whether request delete/revokes, if not checks whether provisioning is needed via `shouldProvision`. The `shouldProvision` checks whether `OB` name is present on `OBC` CRD. If exists it returns true. In case of updating `OBC` field this always exist and skips further part of code. Is this intended behavior? 

CCing @yuvalif

### PR DESCRIPTION
**Pull request by thotz**
Adding support to handle Update Events of OBC, currently `AdditionalConfig` needs this change.
Other fields are immutable
Fixes #195
Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>

### CODE REVIEW

**Comment by thotz** [comment]
@BlaineEXE @jeffvance @copejon @leseb PTAL

File: pkg/provisioner/api/provisioner.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ type Provisioner interface {
 	Provision(options *BucketOptions) (*v1alpha1.ObjectBucket, error)
 	// Grant should be implemented to handle access to existing buckets
 	Grant(options *BucketOptions) (*v1alpha1.ObjectBucket, error)
+	// Update should be implemented to handle updation of existing buckets
+	Update(options *BucketOptions) error
```
Comment (jeffvance) [inline pkg/provisioner/api/provisioner.go:47]: I think this is a breaking change in that **all** provisioners will need to add an Update method when they recompile.

File: pkg/provisioner/controller.go [hunk 2]
```diff
@@ -245,8 +246,8 @@ func (c *obcController) syncHandler(key string) error {
 	// Provision New Bucket or Grant Access to Existing Bucket
 	// *******************************************************
 	if !shouldProvision(obc) {
-		log.Info("skipping provision")
-		return nil
+		log.Info("updating OBC")
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:249]: comments on lines 238 and 246 should include that the event could also be update.

File: pkg/provisioner/controller.go [hunk 3]
```diff
@@ -638,3 +639,38 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
 }
+
+func (c *obcController) handleUpdateClaim(key string, obc *v1alpha1.ObjectBucketClaim, class *storagev1.StorageClass) error {
+	ob, err := getObjectBucketFromClaimKey(key, obc, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
+	if err != nil {
+		log.Error(err, "getting ob failed", "key", key)
+		return err
+	}
+	if reflect.DeepEqual(ob.Spec.Endpoint.AdditionalConfigData, obc.Spec.AdditionalConfig) {
+		logD.Info("No updation required")
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:650]: "updation" is not a word (in English). Try "Only additionalConfigData" can be changed. Update ignored."

File: pkg/provisioner/controller.go [hunk 4]
```diff
@@ -638,3 +639,38 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+		return nil
+	} else {
+		logD.Info("Additional config changed, OB updation required")
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:653]: "updation" is not a word.

File: pkg/provisioner/controller.go [hunk 5]
```diff
@@ -638,3 +639,38 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+		bucketName, _ := composeBucketName(obc)
+		options := &api.BucketOptions{
+			ReclaimPolicy:     class.ReclaimPolicy,
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:656]: The storage class could have been deleted and then recreated with the same name. We don't want to potentially change the retainPolicy. Why are any field being set other than additionalConfigData?

File: pkg/provisioner/controller.go [hunk 6]
```diff
@@ -638,3 +639,38 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+	if err != nil {
+		log.Error(err, "getting ob failed", "key", key)
+		return err
+	}
+	if reflect.DeepEqual(ob.Spec.Endpoint.AdditionalConfigData, obc.Spec.AdditionalConfig) {
+		logD.Info("No updation required")
+		return nil
+	} else {
+		logD.Info("Additional config changed, OB updation required")
+		bucketName, _ := composeBucketName(obc)
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:654]: why do we need to re-compose the bucketName?

File: pkg/provisioner/controller.go [hunk 7]
```diff
@@ -638,3 +639,38 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+		options := &api.BucketOptions{
+			ReclaimPolicy:     class.ReclaimPolicy,
+			BucketName:        bucketName,
+			ObjectBucketClaim: obc.DeepCopy(),
+			Parameters:        class.Parameters,
+		}
+		err = c.provisioner.Update(options)
+		if err != nil {
+			log.Error(err, "updation from Provisioner failed")
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:663]: "updation"

File: pkg/provisioner/controller.go [hunk 8]
```diff
@@ -638,3 +639,38 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+			return err
+		}
+		for k, v := range obc.Spec.AdditionalConfig {
+			ob.Spec.Endpoint.AdditionalConfigData[k] = v
+		}
+		ob, err = updateObjectBucket(ob, c.libClientset)
+		if err != nil {
+			log.Error(err, "OB updation failed")
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:671]: "updation"

**Comment by jeffvance** [comment]
There is no error reported if the user tries to update OBC fields other than additionalCongifData.
Also, have you consider additionalConfigData other than quota (which the user is allowed to change). For example, the storage class's parameters field can contain bucket config data and is passed to the provisioner. What happens if the user changes this config data via an updating their obc's additionalConfigData. It seems now that a user is allowed to change _any_ bucket config data.

**Comment by BlaineEXE** [comment]
Are these changes even needed any more given this comment? https://github.com/kube-object-storage/lib-bucket-provisioner/issues/195#issuecomment-742282150

**Comment by thotz** [comment]
> Are these changes even needed any more given this comment? [#195 (comment)](https://github.com/kube-object-storage/lib-bucket-provisioner/issues/195#issuecomment-742282150)

Atleast at that time focus was needed for bucket notifications(decided to go with `label notations` instead of `additionalconfig`), now users have similar request for quota as well https://github.com/rook/rook/issues/7146

File: pkg/provisioner/api/provisioner.go (hunk 1 above)
Comment (thotz) [reply pkg/provisioner/api/provisioner.go:47]: That happens only if they update the latest version of the library right? Can we release note or something in which we can mention about this change

File: pkg/provisioner/controller.go (hunk 5 above)
Comment (thotz) [reply pkg/provisioner/controller.go:656]: I will pass `OBC` instead of options that make more sense, provisioner need to ob info (backend bucket and backend user)

**Comment by thotz** [comment]
> There is no error reported if the user tries to update OBC fields other than additionalCongifData.
> Also, have you consider that the user can mutate additionalConfigData other than quota (which the user is allowed to change). For example, the storage class's parameters field can contain bucket config data and is passed to the provisioner. What happens if the user changes this config data by updating their obc's additionalConfigData. It seems now a user is allowed to change _any_ bucket config data.

@jeffvance : I agree that other fields apart from `additionalConfigData` should not be modified, this field is opaque to the library so it won't able to know changes made to `additionalConfigData`. Sorry I didn't fully understand how `storageclass.Parameter` related to `additionalConfigData`

**Comment by jeffvance** [comment]
> @jeffvance : I agree that other fields apart from `additionalConfigData` should not be modified, this field is opaque to the library so it won't able to know changes made to `additionalConfigData`. Sorry I didn't fully understand how `storageclass.Parameter` related to `additionalConfigData`

1) what happens if the user modifies OBC fields other than additionalConfigData? It looks like there will be no error reported.
2) Re. storage class and parameters- nevermind. I didn't remember that your are passing an options struct  to the Update method, which define the field the Update method sees.  Be careful if change to pass a map to Update because then there's a possibility of collisions with the StorageClass.parameters config map.

File: pkg/provisioner/api/provisioner.go (hunk 1 above)
Comment (jeffvance) [reply pkg/provisioner/api/provisioner.go:47]: Agree

File: pkg/provisioner/controller.go (hunk 5 above)
Comment (jeffvance) [reply pkg/provisioner/controller.go:656]: What is `additionalState`?

Comment (thotz) [reply pkg/provisioner/controller.go:656]: I guess I will pass OB than separate parameters, so this API will resemble to Delete() or Grant(). `Additional State` another string map that stores information created by the backend of provisioner like user/owner of the bucket
```console
Spec:
  Additional State:
    Ceph User:  ceph-user-up2Dphkq
```

**Comment by jeffvance** [comment]
> now users have similar request for quota as well rook/rook#7146
So do we want to also support notification via additionalConfigData?

**Comment by jeffvance** [comment]
From above:
> what happens if the user modifies OBC fields other than additionalConfigData? It looks like there will be no error reported.

This has not been addressed.

@BlaineEXE what are your thoughts?

**Comment by thotz** [comment]
> > at that time focus was needed for bucket notifications (decided to go with label notations instead of additionalconfig)
> 
> So do we want to also support notification via additionalConfigData?

Currently not planned to do so

File: pkg/provisioner/controller.go [hunk 9]
```diff
@@ -99,6 +100,12 @@ func NewController(provisionerName string, provisioner api.Provisioner, clientse
 			if newObc.ObjectMeta.DeletionTimestamp != nil && oldObc.ObjectMeta.DeletionTimestamp != nil {
 				return
 			}
+			if !reflect.DeepEqual(newObc.Spec, oldObc.Spec) {
+				if reflect.DeepEqual(newObc.Spec.AdditionalConfig, oldObc.Spec.AdditionalConfig) {
+					log.Error(nil, "Only Additional Config can be modified for an OBC")
```
Comment (thotz) [inline pkg/provisioner/controller.go:105]: @jeffvance : will this suffice as an error check for modifying other fields apart from `additionalconfig`, so that update happens specifically.

Comment (jeffvance) [reply pkg/provisioner/controller.go:105]: The `if` in [L100](https://github.com/kube-object-storage/lib-bucket-provisioner/pull/213/files#diff-1e346e3430c7a1bec9a1fde398367f33b921671c6c3ced081ad42e4d8e345522R100) is checking that if _both_ the old and new OBCs are being deleted then `return`.
I think you only need to check if the new OBC is being deleted and then return.

Comment (jeffvance) [reply pkg/provisioner/controller.go:105]: > will this suffice as an error check for modifying other fields apart from additionalconfig?

Yes. I think I missed seeing the first DeepEqual on spec.

Comment (BlaineEXE) [reply pkg/provisioner/controller.go:105]: Go error messages are supposed to not start with capital letters. I also suggest rewording for clarity.

```suggestion
					log.Error(nil, "invalid changes to OBC. only additionalConfig can be updated after creation")
```

File: pkg/provisioner/controller.go [hunk 10]
```diff
@@ -638,3 +649,49 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+		for key, value = range ob.Spec.Endpoint.AdditionalConfigData {
+			tmp[key] = value
+			delete(ob.Spec.Endpoint.AdditionalConfigData, key)
+		}
+
+		for key, value = range obc.Spec.AdditionalConfig {
+			ob.Spec.Endpoint.AdditionalConfigData[key] = value
+		}
+
+		err = c.provisioner.Update(ob)
```
Comment (BlaineEXE) [inline pkg/provisioner/controller.go:678]: I spent some time a while ago making sure that the `Create` method was idempotent.

What about the option of not creating a new `Update` method that has to be implemented by users (i.e., Rook) and instead call Create again? Or does that also pose different issues?

File: pkg/provisioner/controller.go (hunk 9 above)
Comment (BlaineEXE) [reply pkg/provisioner/controller.go:105]: I read that this code checks 
```go
if new spec has changed {
  if new additional config is the same as old additional config {
    logError
    return
    // this part seems correct to me
  }
}
```

But what if a user modifies both the spec and the additionalConfig?

I think the check needs to be more like:
```go
// create copy of old spec, and set the new spec's additionalConfig on it
oldSpecWithNewAdditionalConfig := oldObc.Spec.DeepCopy()
oldSpecWithNewAdditinoalConfig.AdditionalConfig = newObc.Spec.AdditionalConfig.DeepCopy()

if !reflect.DeepEqual(newObc.Spec, oldObc.Spec) {
    // new OBC is changed
    if !reflect.DeepEqual(oldSpecWithNewAdditionalConfig, newObc.Spec) {
        // new OBC has changed something other than additionalConfig
        log.Error(nil, "invalid changes to OBC. only additionalConfig can be updated after creation")
        return
    }
}
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

**Comment by jeffvance** [comment]
> But what if a user modifies both the spec and the additionalConfig?

Good catch @BlaineEXE !

File: pkg/provisioner/controller.go [hunk 11]
```diff
@@ -638,3 +656,48 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+	}
+	if reflect.DeepEqual(ob.Spec.Endpoint.AdditionalConfigData, obc.Spec.AdditionalConfig) {
+		return nil
+	} else {
+		logD.Info("Additional config changed, hence updating OB")
+		tmp := make(map[string]string)
+
+		copyAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData, tmp)
+		clearAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData)
+		copyAdditionalConfig(obc.Spec.AdditionalConfig, ob.Spec.Endpoint.AdditionalConfigData)
```
Comment (BlaineEXE) [inline pkg/provisioner/controller.go:674]: What if, instead of calling `clear...` then `copy...` (which is more of a merge by how it is implemented) there was a function to `overwriteAdditionalConfig` that did both clear and copy (merge) steps? I think that will make it less likely in the future that someone forgets to call `clear` before `copy` when they should do so.

After further thought, It might be more clear to read the functions at a glance as below, which removes the need for `clearAdditionalConfig()` entirely and makes it clear what is being copied and where it is being copied into:
```go
tmp := copyAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData) // copy into tmp
// copy into (and overwrite) OB endpoint's additional config
ob.Spec.Endpoint.AdditionalConfigData = copyAdditionalConfig(obc.Spec.AdditionalConfig) 

func copyAdditionalConfig(a map[string]string) map[string]string {
    ret := map[string]string{}
    for k, v := range a {
        ret[k] = v
    }
    return ret
}
```

File: pkg/provisioner/controller.go [hunk 12]
```diff
@@ -638,3 +656,48 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+			clearAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData)
+			copyAdditionalConfig(tmp, ob.Spec.Endpoint.AdditionalConfigData)
+			log.Error(err, "updating from Provisioner failed")
+			return err
+		}
+		ob, err = updateObjectBucket(ob, c.libClientset)
+		if err != nil {
+			log.Error(err, "updating OB failed")
+			return err
+		}
```
Comment (BlaineEXE) [inline pkg/provisioner/controller.go:688]: Is it necessary to update the ObjectBucket if the Update fails? 

From how I read the function up until this point, If the update has failed, I think there are no changes being made to the ObjectBucket. If that is the case, we should probably just not update the bucket at all. But maybe I missed something.

File: pkg/provisioner/controller.go [hunk 13]
```diff
@@ -638,3 +656,48 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+		clearAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData)
+		copyAdditionalConfig(obc.Spec.AdditionalConfig, ob.Spec.Endpoint.AdditionalConfigData)
+
+		err = c.provisioner.Update(ob)
+		if err != nil {
+			// Provisioner returns error, hence reverting values
+			clearAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData)
+			copyAdditionalConfig(tmp, ob.Spec.Endpoint.AdditionalConfigData)
+			log.Error(err, "updating from Provisioner failed")
+			return err
```
Comment (BlaineEXE) [inline pkg/provisioner/controller.go:682]: These steps are only necessary if the OB actually needs its resource updated following a provisioner `Update()` failure, right?

File: pkg/provisioner/controller.go [hunk 14]
```diff
@@ -638,3 +656,48 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+
+func (c *obcController) handleUpdateClaim(key string, obc *v1alpha1.ObjectBucketClaim, class *storagev1.StorageClass) error {
+	ob, err := getObjectBucketFromClaimKey(key, obc, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
+	if err != nil {
+		log.Error(err, "getting ob failed", "key", key)
+		return err
+	}
+	if reflect.DeepEqual(ob.Spec.Endpoint.AdditionalConfigData, obc.Spec.AdditionalConfig) {
+		return nil
+	} else {
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:668]: more "go-like" to omit the `else` since you've returned in the line above.

Comment (BlaineEXE) [reply pkg/provisioner/controller.go:668]: Ah, yes. I think some linters would complain about this too.

File: pkg/provisioner/controller.go [hunk 15]
```diff
@@ -638,3 +656,48 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+		logD.Info("Additional config changed, hence updating OB")
+		tmp := make(map[string]string)
+
+		copyAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData, tmp)
+		clearAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData)
+		copyAdditionalConfig(obc.Spec.AdditionalConfig, ob.Spec.Endpoint.AdditionalConfigData)
+
+		err = c.provisioner.Update(ob)
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:676]: I know an earlier version of this pr passed an options struct to Update and now you're passing the entire OB, with the need to save a tmp copy of additionalConfig, etc. What is the reason for not just passing to Update a single parameter named `additionalConfig` as a map[string]string? Then provisioners don't need to worry about what else may have changed in the OB. It feels safer and simpler, but at the cost of being less future-proof if it's decided to support more OB mutations.

Comment (thotz) [reply pkg/provisioner/controller.go:676]: @jeffvance : If I pass only the `additionalconfig` the provisioned does not much idea about details like which user or bucket(which might be opaque to library and different for each) it is referring to. Easier way to pass `OB` information like `Delete()` or `Revoke()`. Or I need to figure what all must is needed(from my point of view I can only check from Rook)

File: pkg/provisioner/controller.go [hunk 16]
```diff
@@ -638,3 +655,43 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+		log.Error(err, "updating OB failed")
+		// revert to old values
+		overwriteAdditionalConfig(tmp, ob.Spec.Endpoint.AdditionalConfigData)
+		c.provisioner.Update(ob)
+		return err
+	}
+	return nil
+}
+
+func overwriteAdditionalConfig(srcAdditionalConfig map[string]string, destAdditionalConfig map[string]string) {
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:690]: nit:
`func overwriteAdditionalConfig(srcAdditionalConfig,  destAdditionalConfig map[string]string) {`

File: pkg/provisioner/controller.go [hunk 17]
```diff
@@ -638,3 +655,43 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+		log.Error(err, "getting ob failed", "key", key)
+		return err
+	}
+	if reflect.DeepEqual(ob.Spec.Endpoint.AdditionalConfigData, obc.Spec.AdditionalConfig) {
+		return nil
+	}
+	log.Info("Additional config changed, hence updating OB")
+	tmp := make(map[string]string)
+
+	overwriteAdditionalConfig(ob.Spec.Endpoint.AdditionalConfigData, tmp)
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:671]: Suggest delete L669 and here replace with:
`tmp := ob.Spec.Endpoint.AdditionalConfigData`

File: pkg/provisioner/controller.go [hunk 18]
```diff
@@ -638,3 +655,43 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+	overwriteAdditionalConfig(obc.Spec.AdditionalConfig, ob.Spec.Endpoint.AdditionalConfigData)
+
+	err = c.provisioner.Update(ob)
+	if err != nil {
+		log.Error(err, "updating from Provisioner failed")
+		return err
+	}
+	ob, err = updateObjectBucket(ob, c.libClientset)
+	if err != nil {
```
Comment (jeffvance) [inline pkg/provisioner/controller.go:680]: It seems like this is a good place to do a retry inline vs. requeuing on error. It's potentially expensive for the provisioner to update the new additionalConfig, then update back to the old value, then the event is requeued, then provisioner updates to new config again, etc.

File: pkg/provisioner/controller.go [hunk 19]
```diff
@@ -99,6 +100,18 @@ func NewController(provisionerName string, provisioner api.Provisioner, clientse
 			if newObc.ObjectMeta.DeletionTimestamp != nil && oldObc.ObjectMeta.DeletionTimestamp != nil {
 				return
 			}
+
```
[the remark this answers is not in the document]
Comment (thotz) [reply pkg/provisioner/controller.go:103]: `label` and `annotation` is not part of `ObjectBucketClaimSpec` this `Update` API or previous version of `Updatefunc` was not checking that. If needed we can define different API for those ones. And even `Provisioner`(like Rook) can itself(or has authority) to check for labels and annotations. Or is the duty to library inform the `Provisioner` for these changes.

File: pkg/provisioner/controller.go [hunk 20]
```diff
@@ -99,6 +100,18 @@ func NewController(provisionerName string, provisioner api.Provisioner, clientse
...
+			// create copy of old spec, and set the new spec's additionalConfig on it
+			oldSpecWithNewAdditionalConfig := oldObc.Spec.DeepCopy()
+			overwriteAdditionalConfig(newObc.Spec.AdditionalConfig, oldSpecWithNewAdditionalConfig.AdditionalConfig)
+			if !reflect.DeepEqual(newObc.Spec, oldObc.Spec) {
+				// new OBC is changed
+				if !reflect.DeepEqual(*oldSpecWithNewAdditionalConfig, newObc.Spec) {
```
[the remark this answers is not in the document]
Comment (thotz) [reply pkg/provisioner/controller.go:109]: without pointer reference the check was always returning false

File: pkg/provisioner/controller.go [hunk 21]
```diff
@@ -638,3 +655,43 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
 	return ob, nil
 }
+
+func (c *obcController) handleUpdateClaim(key string, obc *v1alpha1.ObjectBucketClaim, class *storagev1.StorageClass) error {
+	ob, err := getObjectBucketFromClaimKey(key, obc, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
+	if err != nil {
+		log.Error(err, "getting ob failed", "key", key)
+		return err
+	}
+	if reflect.DeepEqual(ob.Spec.Endpoint.AdditionalConfigData, obc.Spec.AdditionalConfig) {
```
[the remark this answers is not in the document]
Comment (thotz) [reply pkg/provisioner/controller.go:665]: I guess this check may reductant

File: pkg/provisioner/controller.go (hunk 18 above)
Comment (thotz) [reply pkg/provisioner/controller.go:680]: are suggesting the following
`for i=0; i<3 &&  err !=nil; i++ {
ob, err = updateObjectBucket(ob, c.libClientset)
}

File: pkg/provisioner/controller.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (jeffvance) [reply pkg/provisioner/controller.go:109]: Is it always true now? You want to make sure you are not comparing an address with a spec.

**Comment by jeffvance** [comment]
Typo L133: "waith" -> "wait"

File: pkg/provisioner/controller.go (hunk 18 above)
Comment (jeffvance) [reply pkg/provisioner/controller.go:680]: No. Look at examples of using the apimachinary's _wait_ package; (PollImmediate)[https://github.com/kubernetes/apimachinery/blob/v0.21.0/pkg/util/wait/wait.go#L440]. Eg.
```
   err = wait.PollImmediate(2*time.seconds, 2*time.minutes, func() (done bool, err error) {
             _, err = updateObjectBucket(ob, c.libClientset)
    	     if err != nil {
		     log.Error(err, "updating OB failed, retrying...")
		     return false, err
	     }
             return true, nil
    })

    if err != nil {
    	log.Error(err, "updating OB failed")
	// revert to old values
	overwriteAdditionalConfig(tmp, ob.Spec.Endpoint.AdditionalConfigData)
	err = c.provisioner.Update(ob)
    }
    return err
}
```

File: pkg/provisioner/controller.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (thotz) [reply pkg/provisioner/controller.go:109]: IMO `oldSpecWithNewAdditionalConfig` is pointer than a `Spec`. I have added following function to print `Spec` values
```diff
diff --git a/pkg/provisioner/controller.go b/pkg/provisioner/controller.go
index 68f47e9..ddc4cc7 100644
--- a/pkg/provisioner/controller.go
+++ b/pkg/provisioner/controller.go
@@ -105,6 +105,7 @@ func NewController(provisionerName string, provisioner api.Provisioner, clientse
                        oldSpecWithNewAdditionalConfig := oldObc.Spec.DeepCopy()
                        overwriteAdditionalConfig(newObc.Spec.AdditionalConfig, oldSpecWithNewAdditionalConfig.AdditionalConfig)
                        if !reflect.DeepEqual(newObc.Spec, oldObc.Spec) {
+                               printObcSpec(oldSpecWithNewAdditionalConfig)
                                // new OBC is changed
                                if !reflect.DeepEqual(*oldSpecWithNewAdditionalConfig, newObc.Spec) {
                                        // new OBC has changed something other than additionalConfig
@@ -695,3 +696,13 @@ func overwriteAdditionalConfig(srcAdditionalConfig map[string]string, destAdditi
                destAdditionalConfig[key] = value
        }
 }
+
+func printObcSpec(spec v1alpha1.ObjectBucketClaimSpec) {
+       logD.Info("storageclassName ", spec.StorageClassName)
+       logD.Info("bucketname ", spec.BucketName)
+       logD.Info("generatebucknam", spec.GenerateBucketName)
+       logD.Info("ObjectBucketName", spec.ObjectBucketName)
+       for key, value := range spec.AdditionalConfig {
+               logD.Info("key ", key, "value ", value)
+       }
+}
```
And give me following error :
```console
go build'ing package ./pkg/...
# github.com/kube-object-storage/lib-bucket-provisioner/pkg/provisioner
pkg/provisioner/controller.go:108:17: cannot use oldSpecWithNewAdditionalConfig (type *"github.com/kube-object-storage/lib-bucket-provisioner/pkg/apis/objectbucket.io/v1alpha1".ObjectBucketClaimSpec) as type "github.com/kube-object-storage/lib-bucket-provisioner/pkg/apis/objectbucket.io/v1alpha1".ObjectBucketClaimSpec in argument to printObcSpec
```

File: pkg/provisioner/controller.go [hunk 22]
```diff
@@ -638,3 +648,62 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+	ob, err := getObjectBucketFromClaimKey(key, obc, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
+	if err != nil {
+		log.Error(err, "getting ob failed", "key", key)
+		return err
+	}
+
+	log.Info("Additional config changed, hence updating OB")
+	tmp := make(map[string]string)
+
+	tmp = ob.Spec.Endpoint.AdditionalConfigData
```
Comment (BlaineEXE) [inline pkg/provisioner/controller.go:678]: In Go, maps need to be deep copied. This will only change the pointer, and `tmp` will therefore be overwritten by `overwriteAdditionalConfig`. If updating the object bucket fails, this will not properly reset the value.

See my test here: https://play.golang.org/p/z0BurgSZEM8

Given this potential bug, it would be best to have a unit test that verifies the provisioner is properly reset when updating the OB fails.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/provisioner/controller.go [hunk 23]
```diff
@@ -638,3 +648,62 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
...
+			return false, err
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		log.Error(err, "updating OB failed, reverting provisioner to original value")
+		overwriteAdditionalConfig(tmp, ob.Spec.Endpoint.AdditionalConfigData)
+		err = c.provisioner.Update(ob)
+	}
```
Comment (BlaineEXE) [inline pkg/provisioner/controller.go:698]: I think it's best to keep the `if err != nil` connected visually (no space between). I would separate the provisioner update from the OB update with a newline but not the error checking on updating the OB

```suggestion

	err = wait.PollImmediate(defaultRetryTimeout, defaultRetryBaseInterval, func() (done bool, err error) {
		_, err = updateObjectBucket(ob, c.libClientset)
		if err != nil {
			log.Error(err, "updating OB failed, retrying...")
			return false, err
		}
		return true, nil
	})
	if err != nil {
		log.Error(err, "updating OB failed, reverting provisioner to original value")
		overwriteAdditionalConfig(tmp, ob.Spec.Endpoint.AdditionalConfigData)
		err = c.provisioner.Update(ob)
	}
```

File: pkg/provisioner/controller.go (hunk 22 above)
Comment (jeffvance) [reply pkg/provisioner/controller.go:678]: Yes you are right! Sorry @thotz

File: pkg/provisioner/controller.go [hunk 24]
```diff
@@ -638,3 +648,62 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
 	}
 	return ob, nil
 }
+
+func updateSupported(old, new *v1alpha1.ObjectBucketClaim) bool {
+	// The only field supported for update is obc.spec.additionalConfig
+	if reflect.DeepEqual(new.Spec, old.Spec) {
```
Comment (copejon) [inline pkg/provisioner/controller.go:654]: Is it necessary to compare the whole spec if the only supported field is additionalConfig?  Why not instead check if additionalConfig has changed?

Comment (jeffvance) [reply pkg/provisioner/controller.go:654]: One of the goals is to log an error or warning if the user attempts to change anything in the obc other than `.spec.additionalConfig`.  I guess DeepEqual handles map compares in cases where they keys may be in different order, right?

Comment (copejon) [reply pkg/provisioner/controller.go:654]: Nevermind, I believe it's just to short circuit the call if nothing call a non-spec field was changed.  Makes sense to me.

File: pkg/provisioner/controller.go (hunk 22 above)
Comment (thotz) [reply pkg/provisioner/controller.go]: @BlaineEXE : I agree it is good to have a unit test. Sorry but I don't understand which function I need write the unit test.
Is it for `overwriteAdditionalConfig()` with set of string maps like in ur example. Or for the `handleUpdateClaim()`? 
If it is `handleUpdateClaim()`
First I need to create OBC and then call `handleUpdateClaim()` with that OBC for different test cases?

**Review by jeffvance — COMMENTED** [review]
lgtm

(for future reference, I prefer to see pr changes in separate commits so that I can re-review only the code that has changed since my previous review)

**Comment by thotz** [comment]
> lgtm
> 
> (for future reference, I prefer to see pr changes in separate commits so that I can re-review only the code that has changed since my previous review)

Sure I keep this in mind while sending PR. Sorry for the trouble and thanks for your patience while reviewing the PR

**Comment by jeffvance** [comment]
No worries @thotz - single commit PRs are common, but, IMO, there are advantages to multi-commit PRs with sensible squashing just before merging.

File: pkg/provisioner/controller.go (hunk 22 above)
Comment (BlaineEXE) [reply pkg/provisioner/controller.go]: I think for `handleUpdateClaim` and make sure the unit test coverage report activates the lines below
```go
if err != nil {
		log.Error(err, "updating OB failed, reverting provisioner to original value")
    		overwriteAdditionalConfig(tmp, ob.Spec.Endpoint.AdditionalConfigData)
    		err = c.provisioner.Update(ob)
}
```

**Review by BlaineEXE — APPROVED** [review]
I didn't realize I was the hold-up. Sorry. I think this looks good now.

**Comment by leseb** [comment]
@thotz is Rook going to consume that?

**Comment by thotz** [comment]
Yes I can update the library for Rook
